### PR TITLE
Add stateful `compressor`, `expander`, `compander` combinators to `Protocols.Df`

### PR DESCRIPTION
**Pull request by wyager**

### CODE REVIEW

**Comment by martijnbastiaan** [comment]
Sorry for the delay, I was out sick. Could you:

 * [Rebase on top of `origin/main`](https://git-scm.com/docs/git-rebase)
 * Squash your commit message into something informative
 * Add tests for your functions in [`Tests.Protocols.Df`](https://github.com/clash-lang/clash-protocols/blob/main/tests/Tests/Protocols/Df.hs)

?

**Review by martijnbastiaan — CHANGES_REQUESTED** [review]
Sorry for the delay, I was out sick. Could you:

 * [Rebase on top of `origin/main`](https://git-scm.com/docs/git-rebase)
 * Squash your commit message into something informative
 * Add tests for your functions in [`Tests.Protocols.Df`](https://github.com/clash-lang/clash-protocols/blob/main/tests/Tests/Protocols/Df.hs)

?

**Review by martijnbastiaan — APPROVED** [review]
Thank you! I'll leave this open for a bit in case any of the other maintainers have an opinion.

File: src/Protocols/Df.hs [hunk 1]
```diff
@@ -238,6 +241,84 @@ forceResetSanity = forceResetSanityGeneric
...
+compander ::
+  forall dom i o s.
+  (C.HiddenClockResetEnable dom, C.NFDataX s) =>
+  s ->
+  -- | Return `True` when you're finished with the current input value
+  -- and are ready for the next one.
+  -- Return `Just` to send the produced value off to the right.
+  (s -> i -> (s, Maybe o, Bool)) ->
+  Circuit (Df dom i) (Df dom o)
+compander s0 f = forceResetSanity |> Circuit (C.unbundle . go . C.bundle)
```
Comment (gergoerdi) [inline src/Protocols/Df.hs:309]: Could just use `mealyB` here directly

**Pull request merged**